library/core/src/fmt/mod.rs

```diff
@@ -2776,7 +2776,14 @@ impl Display for char {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> Pointer for *const T {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        pointer_fmt_inner(self.expose_provenance(), f)
+        if core::mem::size_of::<<T as core::ptr::Pointee>::Metadata>() == 0 {
```

Comment:
It seems a bit strange to check the size of the metadata. Could we instead explicitly check that the metadata type is `()`?

Reply:
I tried doing that at first but couldn't figure out how. I tried a bit more now but without success. I can't find any other place in the compiler/libraries that does that either.

Maybe you had some particular technique in mind?

Personally I get peace of mind thinking of this logic as "if the metadata has no size there is nothing of interest to show".

Reply:
If we added a `'static` requirement to `Pointee::Metadata`, then you could compare `TypeId`. I'm not sure if there's any good reason it doesn't have that already.

Reply:
Thanks, that seems to work. I think I can push a commit for that to see if CI passes without disrupting the FCP?

Reply:
`Pointee::Metadata` can't be `'static`, because the metadata of `dyn Trait + 'a` is `DynMetadata<dyn Trait + 'a>` which is not `'static`.

To me it seems safer to drop the `'static` bound from the `type_id` intrinsic rather than potentially causing a trait system unsoundness by adding a `'static` bound that isn't actually satisfied by the compiler-builtin impls.

Reply:
That makes sense. I'll let CI continue to see if it catches it, and then go back to the code I had when the FCP started.

Reply:
You could use a local specialization:

```rust
trait IsUnit {
    fn is_unit() -> bool;
}

impl<T: ?Sized> IsUnit for T {
    #[inline(always)]
    default fn is_unit() -> bool { false }
}

impl IsUnit for () {
    #[inline(always)]
    fn is_unit() -> bool { true }
}
```

(It would be better as an associated const, but I don't think `min_specialization` allows that.)

Reply:
Thanks. I have now
* rebased on master to resolve conflict
* added a comment for future adventurers explaining why `Pointee::Metadata` can't be `'static`
* check for `()` with `IsUnit` instead of `size_of() == 0`